// @react-native-community/cli-config-apple/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "findPbxprojFile", {
  enumerable: true,
  get: function () {
    return _findPbxprojFile.default;
  }
});
Object.defineProperty(exports, "findPodfilePaths", {
  enumerable: true,
  get: function () {
    return _config.findPodfilePaths;
  }
});
Object.defineProperty(exports, "findXcodeProject", {
  enumerable: true,
  get: function () {
    return _findXcodeProject.default;
  }
});
Object.defineProperty(exports, "getDependencyConfig", {
  enumerable: true,
  get: function () {
    return _config.getDependencyConfig;
  }
});
Object.defineProperty(exports, "getProjectConfig", {
  enumerable: true,
  get: function () {
    return _config.getProjectConfig;
  }
});
Object.defineProperty(exports, "installPods", {
  enumerable: true,
  get: function () {
    return _installPods.default;
  }
});
Object.defineProperty(exports, "resolvePods", {
  enumerable: true,
  get: function () {
    return _pods.default;
  }
});
Object.defineProperty(exports, "supportedPlatforms", {
  enumerable: true,
  get: function () {
    return _supportedPlatforms.supportedPlatforms;
  }
});
var _config = require("./config");
var _installPods = _interopRequireDefault(require("./tools/installPods"));
var _pods = _interopRequireDefault(require("./tools/pods"));
var _findXcodeProject = _interopRequireDefault(require("./config/findXcodeProject"));
var _findPbxprojFile = _interopRequireDefault(require("./config/findPbxprojFile"));
var _supportedPlatforms = require("./config/supportedPlatforms");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
