// @react-navigation/core/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  BaseNavigationContainer: true,
  createNavigationContainerRef: true,
  createNavigatorFactory: true,
  CurrentRenderContext: true,
  findFocusedRoute: true,
  getActionFromState: true,
  getFocusedRouteNameFromRoute: true,
  getPathFromState: true,
  getStateFromPath: true,
  NavigationContainerRefContext: true,
  NavigationContext: true,
  NavigationHelpersContext: true,
  NavigationIndependentTree: true,
  NavigationRouteContext: true,
  PreventRemoveContext: true,
  PreventRemoveProvider: true,
  createComponentForStaticNavigation: true,
  createPathConfigForStaticNavigation: true,
  ThemeContext: true,
  ThemeProvider: true,
  useTheme: true,
  useFocusEffect: true,
  useIsFocused: true,
  useNavigation: true,
  useNavigationBuilder: true,
  useNavigationContainerRef: true,
  useNavigationIndependentTree: true,
  useNavigationState: true,
  usePreventRemove: true,
  usePreventRemoveContext: true,
  useRoute: true,
  validatePathConfig: true
};
Object.defineProperty(exports, "BaseNavigationContainer", {
  enumerable: true,
  get: function () {
    return _BaseNavigationContainer.BaseNavigationContainer;
  }
});
Object.defineProperty(exports, "CurrentRenderContext", {
  enumerable: true,
  get: function () {
    return _CurrentRenderContext.CurrentRenderContext;
  }
});
Object.defineProperty(exports, "NavigationContainerRefContext", {
  enumerable: true,
  get: function () {
    return _NavigationContainerRefContext.NavigationContainerRefContext;
  }
});
Object.defineProperty(exports, "NavigationContext", {
  enumerable: true,
  get: function () {
    return _NavigationContext.NavigationContext;
  }
});
Object.defineProperty(exports, "NavigationHelpersContext", {
  enumerable: true,
  get: function () {
    return _NavigationHelpersContext.NavigationHelpersContext;
  }
});
Object.defineProperty(exports, "NavigationIndependentTree", {
  enumerable: true,
  get: function () {
    return _NavigationIndependentTree.NavigationIndependentTree;
  }
});
Object.defineProperty(exports, "NavigationRouteContext", {
  enumerable: true,
  get: function () {
    return _NavigationRouteContext.NavigationRouteContext;
  }
});
Object.defineProperty(exports, "PreventRemoveContext", {
  enumerable: true,
  get: function () {
    return _PreventRemoveContext.PreventRemoveContext;
  }
});
Object.defineProperty(exports, "PreventRemoveProvider", {
  enumerable: true,
  get: function () {
    return _PreventRemoveProvider.PreventRemoveProvider;
  }
});
Object.defineProperty(exports, "ThemeContext", {
  enumerable: true,
  get: function () {
    return _ThemeContext.ThemeContext;
  }
});
Object.defineProperty(exports, "ThemeProvider", {
  enumerable: true,
  get: function () {
    return _ThemeProvider.ThemeProvider;
  }
});
Object.defineProperty(exports, "createComponentForStaticNavigation", {
  enumerable: true,
  get: function () {
    return _StaticNavigation.createComponentForStaticNavigation;
  }
});
Object.defineProperty(exports, "createNavigationContainerRef", {
  enumerable: true,
  get: function () {
    return _createNavigationContainerRef.createNavigationContainerRef;
  }
});
Object.defineProperty(exports, "createNavigatorFactory", {
  enumerable: true,
  get: function () {
    return _createNavigatorFactory.createNavigatorFactory;
  }
});
Object.defineProperty(exports, "createPathConfigForStaticNavigation", {
  enumerable: true,
  get: function () {
    return _StaticNavigation.createPathConfigForStaticNavigation;
  }
});
Object.defineProperty(exports, "findFocusedRoute", {
  enumerable: true,
  get: function () {
    return _findFocusedRoute.findFocusedRoute;
  }
});
Object.defineProperty(exports, "getActionFromState", {
  enumerable: true,
  get: function () {
    return _getActionFromState.getActionFromState;
  }
});
Object.defineProperty(exports, "getFocusedRouteNameFromRoute", {
  enumerable: true,
  get: function () {
    return _getFocusedRouteNameFromRoute.getFocusedRouteNameFromRoute;
  }
});
Object.defineProperty(exports, "getPathFromState", {
  enumerable: true,
  get: function () {
    return _getPathFromState.getPathFromState;
  }
});
Object.defineProperty(exports, "getStateFromPath", {
  enumerable: true,
  get: function () {
    return _getStateFromPath.getStateFromPath;
  }
});
Object.defineProperty(exports, "useFocusEffect", {
  enumerable: true,
  get: function () {
    return _useFocusEffect.useFocusEffect;
  }
});
Object.defineProperty(exports, "useIsFocused", {
  enumerable: true,
  get: function () {
    return _useIsFocused.useIsFocused;
  }
});
Object.defineProperty(exports, "useNavigation", {
  enumerable: true,
  get: function () {
    return _useNavigation.useNavigation;
  }
});
Object.defineProperty(exports, "useNavigationBuilder", {
  enumerable: true,
  get: function () {
    return _useNavigationBuilder.useNavigationBuilder;
  }
});
Object.defineProperty(exports, "useNavigationContainerRef", {
  enumerable: true,
  get: function () {
    return _useNavigationContainerRef.useNavigationContainerRef;
  }
});
Object.defineProperty(exports, "useNavigationIndependentTree", {
  enumerable: true,
  get: function () {
    return _useNavigationIndependentTree.useNavigationIndependentTree;
  }
});
Object.defineProperty(exports, "useNavigationState", {
  enumerable: true,
  get: function () {
    return _useNavigationState.useNavigationState;
  }
});
Object.defineProperty(exports, "usePreventRemove", {
  enumerable: true,
  get: function () {
    return _usePreventRemove.usePreventRemove;
  }
});
Object.defineProperty(exports, "usePreventRemoveContext", {
  enumerable: true,
  get: function () {
    return _usePreventRemoveContext.usePreventRemoveContext;
  }
});
Object.defineProperty(exports, "useRoute", {
  enumerable: true,
  get: function () {
    return _useRoute.useRoute;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function () {
    return _useTheme.useTheme;
  }
});
Object.defineProperty(exports, "validatePathConfig", {
  enumerable: true,
  get: function () {
    return _validatePathConfig.validatePathConfig;
  }
});
var _BaseNavigationContainer = require("./BaseNavigationContainer.js");
var _createNavigationContainerRef = require("./createNavigationContainerRef.js");
var _createNavigatorFactory = require("./createNavigatorFactory.js");
var _CurrentRenderContext = require("./CurrentRenderContext.js");
var _findFocusedRoute = require("./findFocusedRoute.js");
var _getActionFromState = require("./getActionFromState.js");
var _getFocusedRouteNameFromRoute = require("./getFocusedRouteNameFromRoute.js");
var _getPathFromState = require("./getPathFromState.js");
var _getStateFromPath = require("./getStateFromPath.js");
var _NavigationContainerRefContext = require("./NavigationContainerRefContext.js");
var _NavigationContext = require("./NavigationContext.js");
var _NavigationHelpersContext = require("./NavigationHelpersContext.js");
var _NavigationIndependentTree = require("./NavigationIndependentTree.js");
var _NavigationRouteContext = require("./NavigationRouteContext.js");
var _PreventRemoveContext = require("./PreventRemoveContext.js");
var _PreventRemoveProvider = require("./PreventRemoveProvider.js");
var _StaticNavigation = require("./StaticNavigation.js");
var _ThemeContext = require("./theming/ThemeContext.js");
var _ThemeProvider = require("./theming/ThemeProvider.js");
var _useTheme = require("./theming/useTheme.js");
var _types = require("./types.js");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
var _useFocusEffect = require("./useFocusEffect.js");
var _useIsFocused = require("./useIsFocused.js");
var _useNavigation = require("./useNavigation.js");
var _useNavigationBuilder = require("./useNavigationBuilder.js");
var _useNavigationContainerRef = require("./useNavigationContainerRef.js");
var _useNavigationIndependentTree = require("./useNavigationIndependentTree.js");
var _useNavigationState = require("./useNavigationState.js");
var _usePreventRemove = require("./usePreventRemove.js");
var _usePreventRemoveContext = require("./usePreventRemoveContext.js");
var _useRoute = require("./useRoute.js");
var _validatePathConfig = require("./validatePathConfig.js");
var _routers = require("@react-navigation/routers");
Object.keys(_routers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _routers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _routers[key];
    }
  });
});
